osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/infra/networking/DlmsPushNotificationDecoder.java

```diff
@@ -65,19 +73,156 @@ private void reset() {
+         *  09 06 00 00 19 09 00 FF                                  // Logical name: Push setup schedule
+         *
+         *  0F                                                       // Data-notification
+         *  00 00 00 01                                              // Long-invoke-id-and-priority (can be ignored)
+         *  00                                                       // Date-time (empty)
+         *  02 03                                                    // Data-value: structure with 3 elements
+         *  09 11 45 58 58 58 58 31 32 33 34 35 36 37 38 39 30 31 32 // Equipment id EXXXX123456789012
+         *  09 06 00 01 19 09 00 FF                                  // Logical name: Push setup alarms
+         *  06 00 00 00 02                                           // Alarm register, with Replace battery set
+         *
+         *  Notes:
+         *  - For SMR5 alarms, we get 8 additional addressing bytes in front of the alarm. These bytes can be
+         *  ignored here.
+         *  - To check if the alarm is in DSMR4 or SMR5 format, check the 9th byte (at index 8). If it is 0F, then it
+         *  is SMR5, otherwise it is DSMR4, because the 9th byte in DSMR4 is in the identifier and this should be a
+         *  number or a character, so it can't be ASCII code 0F.
+         * */
+
+        boolean smr5alarm = buffer.getByte(8) == 0x0F;
+
+        LOGGER.info("Decoding state: {}, SMR5 alarm: {}", state, smr5alarm);
+
+        if (smr5alarm) {
+            return decodeSmr5alarm(buffer);
+        } else {
+            switch (state) {
+            case EQUIPMENT_IDENTIFIER:
+                this.decodeEquipmentIdentifier(buffer);
+                return this.setCheckpointAndContinueDecode(ctx, channel, buffer, DecodingState.DATA_OBJECT);
+            case DATA_OBJECT:
+                this.decodeReceivedData(buffer);
+                return this.buildPushNotification();
+            default:
+                throw new UnknownDecodingStateException(String.valueOf(state));
+            }
+        }
+    }
+
+    private Object decodeSmr5alarm(final ChannelBuffer buffer) throws UnrecognizedMessageDataException {
+
+        // Skip addressing, the 0x0F byte and the invoke-id-and-priority bytes
+        buffer.skipBytes(SMR5_NUMBER_OF_BYTES_FOR_ADDRESSING + 1 + SMR5_NUMBER_OF_BYTES_FOR_INVOKE_ID);
+
+        // Datetime is not used, so can be skipped as well
+        skipDateTime(buffer);
+
+        // Next byte should indicate a Structure
+        byte dataTypeByte = buffer.readByte();
+        if (dataTypeByte != STRUCTURE) {
+            throw new UnrecognizedMessageDataException("Expected a structure (0x02), but encountered: " + dataTypeByte);
+        }
+        this.builder.appendByte(dataTypeByte);
+
+        // Next byte should indicate the amount of elements in the structure: 2 or 3
+        byte dataLength = buffer.readByte();
+        if (dataLength != 2 && dataLength != 3) {
```

Comment:
I think this should be an OR statement

Reply:
This is correct. If the length is not 2 or 3, then it throws an exception.

Reply:
The boolean expression `dataLength != 2 || dataLength != 3` would always be true! @coendamen

Reply:
yes, I already resolved this comment. I saw the double negation and mistakenly (intuitively) saw it as an OR and moved on with reading.

Reply:
An equivalent expression would be: `!(dataLength == 2 || dataLength == 3)` following De Morgan's law @coendamen